packages/component/src/component_container.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Provider } from './provider';
+import { Component } from './component';
+
+export class ComponentContainer {
+  private providers = new Map<string, Provider>();
+
+  constructor(private name: string) {}
+
+  /**
+   *
+   * @param component Component being added
+   * @param overwrite When a component with the same name has already been registered,
+   * if overwrite is true: overwrite the existing component with the new component and create a new provider with the new component
+   * if overwrite is false: throw an expection
+   */
+  addComponent(component: Component, overwrite = false): void {
+    let provider = this.getProvider(component.name);
+    if (provider.isComponentSet()) {
+      if (!overwrite) {
+        throw new Error(
+          `Component ${component.name} has already been registered with ${this.name}`
+        );
+      } else {
+        // use the new component to replace the existing component
+        // delete the existing provider from the container
+        this.providers.delete(component.name);
+        // create a new provider
+        provider = this.getProvider(component.name);
+      }
+    }
+
+    provider.setComponent(component);
+  }
+
+  getProvider<T>(name: string): Provider<T> {
```

Comment:
Good point. We actually don't need the generics anymore because individual components define overloads.
so we  just have `getProvider(name: string): Provider<unknown>` here, WDYT?

Reply:
I went ahead and removed the generic from the function declaration.

Reply:
Nice, thanks.

This is perfectly fine already, but I think that if the components defined the `NameServicePairs`, it would be more type safe (you wouldn't be able to call this with any string) and it would have the same complexity in the extending components, since they're declaration merging anyway. They'd just extend the interface instead of overloading this method.

Also described here (kind of): http://go/typescript/return_only_generics#mapped-types

Reply:
Done. PTAL.

Reply:
Looks good, thanks. :slightly_smiling_face: